torch/jit/python_frontend.py

```diff
@@ -0,0 +1,285 @@
+        ast.Or: partial(BinOp, 'or'),
+    }
+
+    cmpop_map = {
+        ast.Eq: partial(BinOp, '=='),
+        ast.NotEq: partial(BinOp, '!='),
+        ast.LtE: partial(BinOp, '<='),
+        ast.Lt: partial(BinOp, '<'),
+        ast.GtE: partial(BinOp, '>='),
+        ast.Gt: partial(BinOp, '>'),
+    }
+
+    @staticmethod
+    def build_Attribute(ctx, expr):
+        # NB: the only attributes we support are for getting methods
+        value = build_expr(ctx, expr.value)
+        return ExprBuilder._MethodRef(value, Ident(value.range(), expr.attr))
+
+    @staticmethod
+    def build_Call(ctx, expr):
+        ref = build_expr(ctx, expr.func, allow_methods=True)
+        assert type(ref) is ExprBuilder._MethodRef
+        args = [build_expr(ctx, py_arg) for py_arg in expr.args]
+        kwargs = [Attribute(Ident(name), build_expr(ctx, value)) for name, value in expr.keywords]
+        return Apply(ref.name, [ref.self] + args, kwargs)
+
+    @staticmethod
+    def build_Name(ctx, expr):
+        assert not expr.id.startswith(_reserved_prefix)
+        r = ctx.make_range(expr.lineno, expr.col_offset, expr.col_offset + len(expr.id))
+        return Var(Ident(r, expr.id))
+
+    @staticmethod
+    def build_BinOp(ctx, expr):
+        op = type(expr.op)
+        if op in ExprBuilder.binop_map:
+            return ExprBuilder.binop_map[op](build_expr(ctx, expr.left), build_expr(ctx, expr.right))
+        else:
+            raise RuntimeError("unsupported binary operator: " + op.__name__)
+
+    @staticmethod
+    def build_UnaryOp(ctx, expr):
+        op = type(expr.op)
+        if op in ExprBuilder.unop_map:
+            constructor, token = ExprBuilder.unop_map[op]
+            r = ctx.make_range(expr.lineno, expr.col_offset, expr.col_offset + len(token))
+            return constructor(r, token, build_expr(ctx, expr.operand))
+        else:
+            raise RuntimeError("unsupported unary operator: " + op.__name__)
+
+    @staticmethod
+    def build_BoolOp(ctx, expr):
+        op = type(expr.op)
+        if op in ExprBuilder.boolop_map:
+            constr = ExprBuilder.boolop_map[op]
+        else:
+            raise RuntimeError("unsupported boolean operator: " + op.__name__)
+        if len(expr.values) < 2:
+            raise AssertionError("expected at least 2 values in BoolOp, but got " + str(len(expr.values)))
+        lhs = build_expr(ctx, expr.values[0])
+        for rhs in expr.values[1:]:
+            lhs = constr(lhs, build_expr(ctx, rhs))
+        return lhs
+
+    @staticmethod
+    def build_IfExp(ctx, expr):
+        return TernaryIf(build_expr(ctx, expr.test),
+                         build_expr(ctx, expr.body),
+                         build_expr(ctx, expr.orelse))
+
+    @staticmethod
+    def build_Compare(ctx, expr):
+        operands = [build_expr(ctx, e) for e in [expr.left] + list(expr.comparators)]
+        # NB: Python allows doing x < y.add_(2) < z, and we need to ensure that y.add_(2)
```

Comment:
Can you change the TreeRef AST to have comparisons of the same form as python, and do this in the IR where it is not an issue?

Reply:
I will change this to a TODO, and do it in a followup PR.